impeller/entity/contents/filters/gaussian_blur_filter_contents.cc

```diff
@@ -0,0 +1,283 @@
+    const Entity& entity,
+    const Matrix& effect_transform) const {
+  if (inputs.empty()) {
+    return {};
+  }
+
+  std::optional<Rect> input_coverage = inputs[0]->GetCoverage(entity);
+  if (!input_coverage.has_value()) {
+    return {};
+  }
+
+  Scalar blur_radius = CalculateBlurRadius(sigma_);
+  Vector3 blur_radii =
+      effect_transform.Basis() * Vector3{blur_radius, blur_radius, 0.0};
+  return input_coverage.value().Expand(Point(blur_radii.x, blur_radii.y));
+}
+
+std::optional<Entity> GaussianBlurFilterContents::RenderFilter(
+    const FilterInput::Vector& inputs,
+    const ContentContext& renderer,
+    const Entity& entity,
+    const Matrix& effect_transform,
+    const Rect& coverage,
+    const std::optional<Rect>& coverage_hint) const {
+  if (inputs.empty()) {
+    return std::nullopt;
+  }
+
+  std::optional<Snapshot> input_snapshot =
+      inputs[0]->GetSnapshot("GaussianBlur", renderer, entity,
+                             /*coverage_limit=*/coverage_hint);
+  if (!input_snapshot.has_value()) {
+    return std::nullopt;
+  }
+
+  if (sigma_ < kEhCloseEnough) {
+    return Entity::FromSnapshot(input_snapshot.value(), entity.GetBlendMode(),
+                                entity.GetClipDepth());  // No blur to render.
+  }
+
+  Scalar blur_radius = CalculateBlurRadius(sigma_);
+  Scalar desired_scale = 1.0 / CalculateScale(blur_radius);
+  Vector2 downsample =
+      CalculateIntegerScale(desired_scale, input_snapshot->texture->GetSize());
+
+  // TODO(gaaclarke): This isn't taking into account the blur radius to expand
+  //                  the rendered size, so blurred objects are clipped. In
+  //                  order for that to be implemented correctly we'll need to
+  //                  start adjusting the geometry coordinates in the downsample
+  //                  step so that there is a border of transparency around it
+  //                  before the blur steps.
+  ISize subpass_size =
+      ISize(input_snapshot->texture->GetSize().width / downsample.x,
+            input_snapshot->texture->GetSize().height / downsample.y);
+
+  Quad uvs =
+      CalculateUVs(inputs[0], entity, input_snapshot->texture->GetSize());
+
+  std::shared_ptr<Texture> pass1_out_texture = MakeDownsampleSubpass(
+      renderer, input_snapshot->texture, input_snapshot->sampler_descriptor,
+      uvs, subpass_size);
+
+  Size pass1_pixel_size(1.0 / pass1_out_texture->GetSize().width,
+                        1.0 / pass1_out_texture->GetSize().height);
+
+  std::shared_ptr<Texture> pass2_out_texture = MakeBlurSubpass(
+      renderer, pass1_out_texture, input_snapshot->sampler_descriptor,
+      GaussianBlurFragmentShader::BlurInfo{
+          .blur_uv_offset = Point(0.0, pass1_pixel_size.height),
+          .blur_sigma = sigma_ / downsample.y,
+          .blur_radius = blur_radius / downsample.y,
+          .step_size = 1.0,
+      });
+
+  // TODO(gaaclarke): Make this pass reuse the texture from pass1.
```

Comment:
We should be able to do: 
`Downsample(src: input, dst: texture1) -> Horizontal Blur(src: texture1, dst: texture2) -> Vertical Blur(src: texture2, dst: texture1) -> texture1` 

Right now we are doing: 
`Downsample(src: input, dst: texture1) -> Horizontal Blur(src: texture, dst: texture2) -> Vertical Blur(src: texture2, dst: texture3) -> texture3`.  

I haven't looked into what plumbing we'd need to make this work yet, the ol' ping-pong maneuver.  Potentially it will require different synchronization but I suspect the way we are doing passes now it won't be needed.  I haven't looked into it yet.

Reply:
FWIW I've thought about building this into the render target cache itself, but I don't have a good way to enforce that we don't create self dependencies. Perhaps if we deferred actual texture creation (from @chinmaygarde 's suggestion) and just used descriptors, it would be easier to apply these sorts of optimizations generally.

Reply:
Ah, yeah that makes sense. The first thing that popped into my mind was that you were planning to also sample from the first texture for something. 😆